src/transformStateWithClones.js

```diff
@@ -6,8 +6,45 @@
  *
  * @return {Object[]}
  */
+/**
+ * @param {Object} state
+ * @param {Object[]} actions
+ */
 function transformStateWithClones(state, actions) {
-  // write code here
+  return actions.reduce((history, action, index) => {
+    let newState = index === 0 ? { ...state } : { ...history[index - 1] };
+
+    if (action.type && action.type !== '') {
+      switch (action.type) {
+        case 'clear':
+          newState = {};
+          break;
+        case 'addProperties':
+          if (!action.extraData) {
+            throw new Error('Missing extraData type in addProperties action.');
+          }
+          Object.assign(newState, action.extraData);
+          break;
+        case 'removeProperties':
+          if (!action.keysToRemove) {
+            throw new Error(
+              'Missing keysToRemove type in removeProperties action.',
+            );
+          }
+
+          if (Array.isArray(action.keysToRemove)) {
+            action.keysToRemove.forEach((key) => delete newState[key]);
+          }
+          break;
+        default:
+          throw new Error(`Unsupported action type: ${action.type}`);
+      }
+    } else {
+      throw new Error('Action type must be provided and non-empty!');
+    }
+
+    return [...history, { ...newState }];
```

Comment:
This line here dismantles the history array and the newState object, here are some hints:
- *history* is the returning result, or accumulator, of your reduce. It doesn't need to be reassigned in this case, you could use the same array and use a `.push` function;
- *newState* is being dismantled into a new object that is being passed to the new history array. This is also not necessary, as this is the last line of your return, and newState will be recreated in the next reduce iteration, there's no need to also dismantle it here.

With this in mind, this line could be replaced for:
```
  history.push(newState);   //this adds newState to the history array without recreating any of them
  
  return history;    //this passes history to the next iteration of your reduce method without recreating it
```